src/core/ext/transport/chttp2/server/chttp2_server.cc

```diff
@@ -1076,12 +1080,25 @@ void grpc_server_add_channel_from_fd(grpc_server* server, int fd,
 
   grpc_core::ChannelArgs server_args = core_server->channel_args();
   std::string name = absl::StrCat("fd:", fd);
+  // The fd is going to be registered with the gRPC poller to be asynchronously
+  // notified for read/write/error events. It must be marked as NonBlocking
+  // to ensure that the poller/gRPC executor threads do not get blocked on
+  // the socket when attempting to read/write from it.
+  grpc_core::PosixSocketWrapper sock(fd);
+  (void)sock.SetSocketNonBlocking(true);
```

Comment:
I'm curious, why is it the transport's responsibility to set up the socket correctly? Shouldn't that be the caller's responsibility?

Reply:
No I think the API assumes gRPC takes full control over the fd and is responsible for closing it
Unless these socket options are set, gRPC cannot work with the fd though